woocommerce/woocommerce
# fix: opt in checkbox to respect the user choice when previously opted out

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR took on a life of its own and turned out to be a fair bit more complex to implement than originally intended, as it seems that `woocommerce_allow_tracking` is set to 'no' by default when WooCommerce is installed.

This made it necessary to implement a way to determine if the user has completed the Opt-in step (and/or the entire Core Profiler) previously. This was implemented by storing an object within `woocommerce_onboarding_profile_progress` under the key: `core_profiler_completed_steps`. 

The reason it has a second level of `core_profiler_completed_steps` instead of directly storing the steps is because it leaves open the possibility of storing different step completions for different onboarding flows, e.g for mobile

This object looks like: 

```
{
	"core_profiler_completed_steps": {
		"intro-opt-in": {
			"completed_at": "2024-11-12T06:50:21Z"
		},
		"user-profile": {
			"completed_at": "2024-11-12T06:50:35Z"
		},
		"business-info": {
			"completed_at": "2024-11-12T06:47:24Z"
		}
	}
}
```

Part of implementing this involved adding new endpoints for setting steps to complete, and while implementing this I realised that there was a race condition caused by two API calls both writing to the `woocommerce_onboarding_profile` option. Thus I've implemented ~a mutex to ensure that there's no data loss during a race condition.~ it in a different option instead of within `woocommerce_onboarding_profile`. 

(I could not get the mutex to work reliably due to WP object caching on the host side)



Closes https://github.com/woocommerce/woocommerce/issues/49895.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

There should be 3 scenarios to test:
1. Site is new and hence user has not completed the Core Profiler or Opted In => Checkbox should be checked by default
2. User has previously opted out, so `woocommerce_allow_tracking` should be 'no' => Checkbox should not be checked
3. User has previously opted in, so `woocommerce_allow_tracking` should be 'yes' => Checkbox should be checked

To test this, 
1. Set up WooCommerce along with WC Beta Tester plugin for easy access to change the `woocommerce_allow_tracking` option
2. Go through the Core Profiler onboarding, and on the first page there is a checkbox indicating usage tracking opt-in.
3. Assert the above behaviours hold. You can delete the options (`woocommerce_allow_tracking` and `woocommerce_onboarding_profile` using the WC admin test helper.
4. These behaviours should hold even through a refresh or using the forward/back browser buttons.

<img width="983" alt="image" src="https://github.com/user-attachments/assets/a60d099c-a177-430f-ba41-76f7ca1fdd10">





### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
